src/Stratis.Features.FederatedPeg/Wallet/MultiSigTransactions.cs

```diff
@@ -0,0 +1,187 @@
+        {
+            lock (this.lockObject)
+            {
+                bool res = this.transactionDict.Remove(transactionData.Key);
+
+                if (this.spendableTransactionDict.ContainsKey(transactionData.Key))
+                    this.spendableTransactionDict.Remove(transactionData.Key);
+
+                this.TryRemoveWithdrawal(transactionData);
+
+                transactionData.Subscribe(null);
+
+                return res;
+            }
+        }
+
+        public IEnumerator<TransactionData> GetEnumerator()
+        {
+            lock (this.lockObject)
+            {
+                return this.transactionDict.Values.ToList().GetEnumerator();
+            }
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return this.transactionDict.Values.ToList().GetEnumerator();
+        }
+
+        public void Clear()
+        {
+            lock (this.lockObject)
+            {
+                this.transactionDict.Clear();
+                this.spendableTransactionDict.Clear();
+                this.withdrawalsByDepositDict.Clear();
+            }
+        }
+
+        public void BeforeSpendingDetailsChanged(TransactionData transactionData)
+        {
+            lock (this.lockObject)
+            {
+                if (transactionData.IsSpendable())
+                    this.spendableTransactionDict.Remove(transactionData.Key);
+
+                this.TryRemoveWithdrawal(transactionData);
+            }
+        }
+
+        public void AfterSpendingDetailsChanged(TransactionData transactionData)
+        {
+            lock (this.lockObject)
+            {
+                if (transactionData.IsSpendable())
+                    this.spendableTransactionDict.Add(transactionData.Key, transactionData);
+
+                this.TryAddWithdrawal(transactionData);
+            }
+        }
+
+        public (uint256 depositId, List<TransactionData>)[] GetSpendingTransactionsByDepositId(uint256 depositId = null)
+        {
+            lock (this.lockObject)
+            {
+                if (depositId != null)
+                {
+                    if (this.withdrawalsByDepositDict.TryGetValue(depositId, out List<TransactionData> txList))
+                        return new[] { (depositId, txList) };
+                    else
+                        return new[] { (depositId, new List<TransactionData>()) };
+                }
+
+                return this.withdrawalsByDepositDict.Select(kv => (kv.Key, kv.Value)).ToArray();
+            }
+        }
+
+        /// <summary>
+        /// List all spendable transactions in a multisig address.
+        /// </summary>
+        /// <returns></returns>
+        public TransactionData[] UnspentTransactions()
+        {
+            lock (this.lockObject)
+            {
+                return this.spendableTransactionDict.Select(kv => kv.Value).ToArray();
+            }
+        }
+
+        public bool Contains(TransactionData item)
```

Comment:
Unused. Probably not an issue but maybe it could be IEnumerable?

Reply:
`CopyTo` is used internally when doing a `ToList` such as here;

```
![image](https://user-images.githubusercontent.com/29645989/58528364-965d4900-8219-11e9-8096-8cbbd3224928.png)
```

Reply:
Making it `IEnumeable` caused Json failures:
`One or more errors occurred. (Cannot create and populate list type Stratis.Features.FederatedPeg.Wallet.MultiSigTransactions. Path 'multiSigAddress.transactions', line 1, position 2755.)`

Reply:
Ah I didn't realise that would be an issue, but it still seems like we are getting .Transactions very often to loop over or use Linq on, when the point is to minimise these scans

Reply:
Yes, we should continue to identify places to optimize but also weigh up the cost of adding more dictionaries (e.g. a new `NONspendableTransactionsDict`) with the benefit achieved. For example it may be that the dict would contain 95% of all transactions and thus not provide much speedup benefit.